### PR TITLE
refactor: perf: dedupe logic, inf checks in predict.py

### PR DESCRIPTION
**Pull request by NickCrews**
1. This factors out this logic so that they will stay in sync.
I wet with slightly less verbose names as well
2. Added type hints
3. Possibly improved performance: Before, when we
templated `bayes_factor_expr` into `match_prob_expr`, then
the check for infinity would happen twice, nestedly.
This doesn't affect the behavior,
but it may have had a performance
degradation.
Also, I don't think the CASE was needed for the BF expr,
since SQL engines should be able to handle propagating inf through
multiplication. I think it only is need to fixup the match_probability expression.
So I removed that in a followup commit as well.
4. Use list comprehension instead of map() for readibility

### Type of PR

- [ ] BUG
- [ ] FEAT
- [x] MAINT
- [ ] DOC

### PR Checklist

- [x] Added documentation for changes
- [x] Added feature to example notebooks or tutorial (if appropriate)
- [ you tell me if you think existing coverage is adequate] Added tests (if appropriate) 
- [x] Made changes based off the latest version of Splink
- [x] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint.html)

### CODE REVIEW

File: splink/predict.py [hunk 1]
```diff
@@ -182,3 +141,24 @@ def predict_from_agreement_pattern_counts_sqls(
...
+    if prior == 1.0:
+        bf_expr = sql_infinity_expr
+        match_prob_expr = "1.0"
+        return bf_expr, match_prob_expr
+
+    bf_prior = prob_to_bayes_factor(prior)
+    bf_raw = f"cast({bf_prior} as float8) * " + " * ".join(bf_terms)
+    mp_raw = f"({bf_raw})/(1+({bf_raw}))"
+
+    # if any BF is Infinity then we need to adjust expression,
```
Comment (NickCrews) [inline splink/predict.py:159]: Are we sure this is really true? SQL engines can't figure out that `inf * 5 == inf`?

Comment (NickCrews) [reply splink/predict.py]: Or I guess that's not what this is catching. This is to fix up `match_prob_expr`. So I don't think this CASE bit is unneeded for the `bf_expr`. I removed this in the followup commit. Should improve performance?

File: splink/predict.py [hunk 2]
```diff
@@ -39,37 +39,17 @@ def predict_from_comparison_vectors_sqls(
...
     select_cols_expr = ",".join(select_cols)
-    mult = []
+    bf_terms = []
     for cc in settings_obj.comparisons:
-        mult.extend(cc._match_weight_columns_to_multiply)
+        bf_terms.extend(cc._match_weight_columns_to_multiply)
 
-    probability_two_random_records_match = (
-        settings_obj._probability_two_random_records_match
+    prior = settings_obj._probability_two_random_records_match
```
Comment (ThomasHepworth) [inline splink/predict.py:46]: @RobinL - do you care if this is changed from `probability_two_random_records_match` -> `prior`?

Imo it doesn't impact readability, but you may have a different view.

**Comment by ThomasHepworth** [comment]
Thanks so much for this Nick, this is great!

I'll approve and merge once we've heard back from Robin.

File: splink/predict.py (hunk 2 above)
Comment (RobinL) [reply splink/predict.py:46]: Yeah looks fine to me

**Review by ThomasHepworth — APPROVED** [review]
Thanks as always Nick

**Pull request merged**